File: Solutions/1584. Min Cost to Connect All Points/1584. Min Cost to Connect All Points.cpp
// c++ Prim's with priority queue(min priority queue) implementation   2021.10.20 mod
class Solution {
public:
    int minCostConnectPoints(vector<vector<int>>& points) {
        int res=0,connected=0,n=points.size(),from=0;
        vector<bool> visited(n);
        priority_queue<pair<int,int>,vector<pair<int,int>>,greater<pair<int,int>>> pq;
        while(++connected<n){
            visited[from]=true;
            for(int i=0;i<n;++i){
                if(!visited[i]){
                    pq.push({abs(points[from][0]-points[i][0])+abs(points[from][1]-points[i][1]),i});
                }
            }
            while(visited[pq.top().second]){
                pq.pop();
            }
            res+=pq.top().first;
            from=pq.top().second;
            pq.pop();
        }
        return res;
    }
};

//Kruskal  ,min priority_queue implementation  2021.10.20
#define ar3 array<int,3>
class Solution {
public:
    int find(vector<int>& ds,int i){
        return ds[i]<0?i:ds[i]=find(ds,ds[i]);
    }
    int minCostConnectPoints(vector<vector<int>>& pts) {
        int n=pts.size(),res=0;
        vector<int> ds(n,-1);
        priority_queue<ar3,vector<ar3>,greater<ar3>> pq;
        for(int i=0;i<n-1;++i){
            for(int j=i+1;j<n;++j){
                pq.push({abs(pts[i][0]-pts[j][0])+abs(pts[i][1]-pts[j][1]),i,j});
            }
        }
        while(pq.size()){
            auto [dis,i,j]=pq.top();pq.pop();
            i=find(ds,i),j=find(ds,j);
            if(i!=j){
                res+=dis;
                if(ds[i]>ds[j])
                    swap(i,j);
                ds[i]+=ds[j];
                ds[j]=i;
                if(ds[find(ds,1)]==-n)
                    break;
            }
        }
        return res;
    }
};


//Kruskal  ,min heap implementation(fetch from back,compare to min priority queue fetch from top())
typedef array<int,3> ar;
class Solution {
public:
    int find(vector<int>& ds,int i){
        return ds[i]<0?i:ds[i]=find(ds,ds[i]);
    }
    int minCostConnectPoints(vector<vector<int>>& pts) {
        int res=0,n=pts.size();
        vector<int> ds(n,-1);
        vector<ar> arr;
        for(auto i=0;i<n-1;++i)
            for(auto j=i+1;j<n;++j)
                arr.push_back({abs(pts[i][0]-pts[j][0])+abs(pts[i][1]-pts[j][1]),i,j});
        make_heap(arr.begin(),arr.end(),greater<ar>());
        while(!arr.empty()){
            pop_heap(arr.begin(),arr.end(),greater<ar>());
            auto [dist,i,j]=arr.back();
            arr.pop_back();
            i=find(ds,i),j=find(ds,j);
            if(i!=j){
                res+=dist;
                ds[i]+=ds[j];
                ds[j]=i;
                if(ds[i]==-n)
                    break;
            }
        }
        return res;
    }
};



//Kruskal v2 much slower,text book union
typedef array<int,3> ar;
class Union{
    int components;
    vector<int> component;
public:
    Union(int n){
        components=n;
        for(int i=0;i<=n;++i)
            component.push_back(i);
    };
    int Find(int a){
        if(component[a]!=a)
            component[a]=Find(component[a]);
        return component[a];
    }
    bool Unite(int a,int b){
        if(Find(a)==Find(b))
            return false;
        component[Find(a)]=b;
        --components;
        return true;
    }
    bool United(){
        return components==1;
    }
};
class Solution {
public:
    int minCostConnectPoints(vector<vector<int>>& points) {
        int res=0,n=points.size();
        vector<ar> arr;
        for(int i=0;i<n-1;++i)
            for(int j=0;j<n;++j)
                arr.push_back({abs(points[i][0]-points[j][0])+abs(points[i][1]-points[j][1]),i,j});
        make_heap(arr.begin(),arr.end(),greater<ar>());
        Union uf(n);
        while(!arr.empty()&&!uf.United()){
            pop_heap(arr.begin(),arr.end(),greater<ar>());
            auto [dist,i,j]=arr.back();
            arr.pop_back();
            if(uf.Unite(i,j))
                res+=dist;
        }
        return res;
    }
};









//optimal Prim's for Complete Graph
class Solution {
public:
    int minCostConnectPoints(vector<vector<int>>& points) {
        int n=points.size(),res=0,i=0,connected=0;
        vector<int> min_d(n,4000000);          //constraint:-10^6 <= xi, yi <= 10^6  thus initialization must >= 4*10^6
        while(++connected<n){
            min_d[i]=INT_MAX;
            int min_j=i;
            for(int j=0;j<n;++j){
                if(min_d[j]!=INT_MAX){
                min_d[j]=min(min_d[j],abs(points[i][0] - points[j][0]) + abs(points[i][1] - points[j][1]));
                min_j = min_d[j] < min_d[min_j] ? j : min_j;
				}
            }
            res+=min_d[min_j];
            i=min_j;
        }
        return res;
    }
};

File: Solutions/1584. Min Cost to Connect All Points/1584. Min Cost to Connect All Points.go
// Prim's with priority queue (min heap implementation)
type MinHeap [][2]int
func (h MinHeap) Len() int{return len(h)}
func (h MinHeap) Less(i,j int) bool{return h[i][0]<h[j][0]}
func (h MinHeap) Swap(i,j int) {h[i],h[j]=h[j],h[i]}
func (h *MinHeap) Push(i interface{}) {
    *h=append(*h,i.([2]int))
}
func (h *MinHeap) Pop() interface{} {
    v:=(*h)[len(*h)-1]
    *h=(*h)[:len(*h)-1]
    return v
}
func (h MinHeap) Peek() interface{} {
    return h[0]
}

func minCostConnectPoints(points [][]int) int {
    res,n,from,connected:=0,len(points),0,1
    pq:=&MinHeap{}
    visited:=make([]bool,n)
    heap.Init(pq)
    for connected<n{
        visited[from]=true
        for i:=0;i<n;i+=1{
            if visited[i]==false{
                dist:=Abs(points[i][0]-points[from][0])+Abs(points[i][1]-points[from][1])
                heap.Push(pq,[2]int{dist,i})
            }
        }
        for visited[pq.Peek().([2]int)[1]]==true{
            heap.Pop(pq)
        }
        res+=pq.Peek().([2]int)[0]
        from=pq.Peek().([2]int)[1]
        connected+=1
    }
    return res
}

func Abs(x int) int {
    if x<0{
        return -x
    }
    return x
}










//Kruskal  ,min heap implementation
type MinHeap [][3]int
func (h MinHeap) Len() int{return len(h)}
func (h MinHeap) Less(i,j int) bool{return h[i][0]<h[j][0]}
func (h MinHeap) Swap(i,j int) {h[i],h[j]=h[j],h[i]}
func (h *MinHeap) Push(i interface{}) {
    *h=append(*h,i.([3]int))
}
func (h *MinHeap) Pop() interface{} {
    v:=(*h)[len(*h)-1]
    *h=(*h)[:len(*h)-1]
    return v
}
func (h MinHeap) Peek() interface{} {
    return h[0]
}

func minCostConnectPoints(p [][]int) int {
    n,res:=len(p),0
    ds:=make([]int,n)
    for i:=0;i<n;i+=1{
        ds[i]=-1
    }
    pq:=&MinHeap{}
    heap.Init(pq)
    for i:=0;i<n;i+=1{
        for j:=i+1;j<n;j+=1{
            dist:=Abs(p[i][0]-p[j][0])+Abs(p[i][1]-p[j][1])
            heap.Push(pq,[3]int{dist,i,j})
            //fmt.Println(dist)
        }
    }
    for pq.Len()>0{
        node:=heap.Pop(pq).([3]int)
        node[1]=find(&ds,node[1])
        node[2]=find(&ds,node[2])
        if node[1]!=node[2]{
            res+=node[0]
            //fmt.Print(" ",res)
            ds[node[1]]+=ds[node[2]]
            ds[node[2]]=node[1]
            if ds[node[1]]==-n{
                break
            }
        }
    }
    return res
}

func find(ds *[]int,i int)int{
    if (*ds)[i]<0{
        return i
    }
    (*ds)[i]=find(ds,(*ds)[i])
    return (*ds)[i]
}

func Abs(x int)int{
    if x<0{
        return -x
    }
    return x
}









//optimal Prim's for Complete Graph
func minCostConnectPoints(p [][]int) int {
    n,res,i,connected:=len(p),0,0,1
    min_d:=make([]int,n)
    for i:=0;i<n;i+=1{
        min_d[i]=4000000
    }
    for connected<n{
        min_d[i]=math.MaxInt32
        min_j:=i
        for j:=0;j<n;j+=1{
            if min_d[j]!=math.MaxInt32{
                dist:=Abs(p[i][0]-p[j][0])+Abs(p[i][1]-p[j][1])
                min_d[j]=min(dist,min_d[j])
                if min_d[j]<min_d[min_j]{
                    min_j=j
                }
            }
        }
        res+=min_d[min_j]
        i=min_j
        connected+=1
    }
    return res
}

func Abs(x int)int{
    if x<0{
        return -x
    }
    return x
}

func min(i,j int)int{
    if(i<j){
        return i
    }
    return j
}